### ISSUE TITLE
When passed `NavigationMode.directional` to MediaQuery, `NavigationMode.traditional` is retuned

### ISSUE THREAD

**Issue by predatorx7** [flutter/flutter#87041]
<!-- Thank you for using Flutter!

     If you are looking for support, please check out our documentation
     or consider asking a question on Stack Overflow:
      * https://flutter.dev/
      * https://api.flutter.dev/
      * https://stackoverflow.com/questions/tagged/flutter?sort=frequent

     If you have found a bug or if our documentation doesn't have an answer
     to what you're looking for, then fill out the template below. Please read
     our guide to filing a bug first: https://flutter.dev/docs/resources/bug-reports
-->

## Steps to Reproduce

<!-- You must include full steps to reproduce so that we can reproduce the problem. -->

1. Run `flutter create sample`.
2. Update the files as follows: ... <!-- include every file that is different from the template app! -->
```dart
MaterialApp(
  theme: AppTheme.data,
  builder: (context, child) {
    final _isLeanback =
        /// Returns true if app is running on Android TV using cached value from android platform channel
        dependencies.actionsAndShortcuts?.isLeanback ?? false;
    if (child == null || !_isLeanback) return child ?? const SizedBox();

    return MediaQuery(
      data: MediaQuery.of(context).copyWith(
       /// This is where I set directional navigation for TV's Dpad remote controller
        navigationMode: NavigationMode.directional,
      ),
      child: child,
    );
  },
)
```
3. Check the value of [MediaQuery.navigationMode] in some child widget using the below code: 
```dart
final _n = MediaQuery.maybeOf(context)?.navigationMode;

log.d('Navigation Mode: $_n');
```

**Expected results:** <!-- what did you want to see? -->
NavigationMode.directional

**Actual results:** <!-- what did you see? -->
NavigationMode.traditional

<details>
  <summary>Output of running `flutter doctor -v`</summary>
<!-- Finally, paste the output of running `flutter doctor -v` here. -->

```                                                                                                                                                                                                                   
[✓] Flutter (Channel stable, 2.2.3, on Linux, locale en_IN.UTF-8)
    • Flutter version 2.2.3 at /home/red/flutter_sdk/versions/2.x
    • Framework revision f4abaa0735 (4 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /home/red/devel/android-sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /home/red/devel/android-sdk
    • Java binary at: /usr/lib/jvm/java-11/bin/java
    • Java version OpenJDK Runtime Environment 18.9 (build 11.0.11+9)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 12.0.0 (Fedora 12.0.0-2.fc34)
    • cmake version 3.20.5
    • ninja version 1.10.2
    • pkg-config version 1.7.3

[✓] Android Studio (version 4.2)
    • Android Studio at /var/lib/snapd/snap/android-studio/113/android-studio
    • Flutter plugin version 58.0.1
    • Dart plugin version 202.8531
    • Java version OpenJDK Runtime Environment (build 11.0.8+0-b944-P17168821)

[!] Android Studio
    • Android Studio at /snap/bin/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ android-studio-dir = /snap/bin/android-studio
    ✗ Android Studio not found at /snap/bin/android-studio
    • Try updating or re-installing Android Studio.
    • Consider removing your android-studio-dir setting by running:
      flutter config --android-studio-dir=

[✓] IntelliJ IDEA Ultimate Edition (version 2021.1)
    • IntelliJ at /var/lib/snapd/snap/intellij-idea-ultimate/298
    • Flutter plugin version 56.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.58.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.24.0

[✓] Connected device (3 available)
    • AFTT (mobile)   • 192.168.1.245:5555 • android-arm    • Android 5.1.1 (API 22)
    • Linux (desktop) • linux              • linux-x64      • Linux
    • Chrome (web)    • chrome             • web-javascript • Google Chrome 92.0.4515.107

! Doctor found issues in 1 category.
```

</details>

## More

My intention was to use actions from a TV controller report without making many changes in widgets. Using the default MaterialApp configs (`actions` & `shortcuts`) with 
```
shortcuts: <LogicalKeySet, Intent>{
      ... WidgetsApp.defaultShortcuts,
      LogicalKeySet(LogicalKeyboardKey.select): const ActivateIntent(),
},
``` 
was causing center or d-pad buttons to not work on some TV remotes for me. I created a platform package to get device configuration and used that configuration to provide custom shortcuts & actions for MaterialApp which did work. Later, I found out  about `navigationMode` can be used for dpad navigations and that `MediaQueryData.fromWindow` is setting `navigationMode = NavigationMode.traditional`. [The documentations mentioned that  `NavigationMode.directional` is for TV dpads](https://github.com/flutter/flutter/blob/184e5871beea4a9b1cbd65dcc57118a7d97285f3/packages/flutter/lib/src/widgets/media_query.dart#L357). As I could not find a parameter in MaterialApp to set navigationMode, I wrote in the `MaterialApp.builder`:
```
data: MediaQuery.of(context).copyWith(
       /// This is where I set directional navigation for TV's Dpad remote controller
        navigationMode: NavigationMode.directional,
      ),
```

**Comment by hndrbrm** [comment]
I try to make one minimalist code:
<details>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Builder(
          builder: (BuildContext context) {
            return MediaQuery(
              data: MediaQuery.of(context).copyWith(
                navigationMode: NavigationMode.directional,
              ),
              child: Builder(
                builder: (BuildContext context) {
                  return Center(
                    child: Text(
                      MediaQuery.maybeOf(context)!.navigationMode.toString(),
                    ),
                  );
                },
              ),
            );
          }
        ),
      ),
    );
  }
}
```
</details>

And i dont see the issue, it works.
It shows `Navigational.directional`

Here is running on [dartpad.dev](https://dartpad.dev/?id=e96dc221edc35c8c4405e3d01b083cc2&null_safety=true)

**Comment by hndrbrm** [comment]
Is it possible that the ```return MediaQuery...``` in the example above doesn't executed.

**Comment by hndrbrm** [comment]
> ## Steps to Reproduce
> 
>     1. Run `flutter create sample`.
> 
>     2. Update the files as follows: ...
> 
> 
> ```dart
> MaterialApp(
>   theme: AppTheme.data,
>   builder: (context, child) {
>     final _isLeanback =
>         /// Returns true if app is running on Android TV using cached value from android platform channel
>         dependencies.actionsAndShortcuts?.isLeanback ?? false;
>     if (child == null || !_isLeanback) return child ?? const SizedBox();
> 
>     return MediaQuery(
>       data: MediaQuery.of(context).copyWith(
>        /// This is where I set directional navigation for TV's Dpad remote controller
>         navigationMode: NavigationMode.directional,
>       ),
>       child: child,
>     );
>   },
> )
> ```
> 
>     1. Check the value of [MediaQuery.navigationMode] in some child widget using the below code:
> 
> 
> ```dart
> final _n = MediaQuery.maybeOf(context)?.navigationMode;
> 
> log.d('Navigation Mode: $_n');
> ```
> 
> **Expected results:**
> NavigationMode.directional
> 
> **Actual results:**
> NavigationMode.traditional
> Output of running `flutter doctor -v`
> ## More
> 
> My intention was to use actions from a TV controller report without making many changes in widgets. Using the default MaterialApp configs (`actions` & `shortcuts`) with
> 
> ```
> shortcuts: <LogicalKeySet, Intent>{
>       ... WidgetsApp.defaultShortcuts,
>       LogicalKeySet(LogicalKeyboardKey.select): const ActivateIntent(),
> },
> ```
> 
> was causing center or d-pad buttons to not work on some TV remotes for me. I created a platform package to get device configuration and used that configuration to provide custom shortcuts & actions for MaterialApp which did work. Later, I found out about `navigationMode` can be used for dpad navigations and that `MediaQueryData.fromWindow` is setting `navigationMode = NavigationMode.traditional`. [The documentations mentioned that `NavigationMode.directional` is for TV dpads](https://github.com/flutter/flutter/blob/184e5871beea4a9b1cbd65dcc57118a7d97285f3/packages/flutter/lib/src/widgets/media_query.dart#L357). As I could not find a parameter in MaterialApp to set navigationMode, I wrote in the `MaterialApp.builder`:
> 
> ```
> data: MediaQuery.of(context).copyWith(
>        /// This is where I set directional navigation for TV's Dpad remote controller
>         navigationMode: NavigationMode.directional,
>       ),
> ```

Is it possible that the return MediaQuery... in the example above doesn't executed.

**Comment by predatorx7** [comment]
@hndrbrm Try this example https://dartpad.dev/?id=e0014767891d15c8d3a5f7c5c6cb3b85&null_safety=true

Thank you, I will use your example as a workaround. Though, I wanted to apply `navigationMode` to all routes, dialogs, overlays.

**Comment by TahaTesser** [comment]
Hi @predatorx7 
Thanks for filing the issue, indeed it prints `NavigationMode.traditional` when passing `NavigationMode.directional`  to the builder in MaterialApp


<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      builder: (context, child) {
        return MediaQuery(
          data: MediaQuery.of(context).copyWith(
            navigationMode: NavigationMode.directional,
          ),
          child: child ?? const SizedBox(),
        );
      },
      home: Scaffold(
        body: Builder(
          builder: (BuildContext context) {
            return Center(
              child: Text(
                MediaQuery.maybeOf(context)!.navigationMode.toString(),
              ),
            );
          },
        ),
      ),
    );
  }
}

``` 

</details>

|  `stable`  | `master` |
| :---: |:---: |
|  ❌  |   ❌  | 

Check `flutter doctor -v` outputs for each channel below

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 2.2.3, on Microsoft Windows [Version 10.0.19043.1110], locale en-US)
    • Flutter version 2.2.3 at C:\Users\Taha\Code\flutter_stable
    • Framework revision f4abaa0735 (4 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\android-sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_SDK_ROOT = C:\Users\Taha\Code\android-sdk
    • Java binary at: C:\Users\Taha\Code\android-studio\jre\bin\java.exe
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31515.178
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] VS Code (version 1.58.2)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.24.0

[✓] Connected device (4 available)
    • sdk google atv x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Windows (desktop)           • windows       • windows-x64    • Microsoft Windows [Version 10.0.19043.1110]
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 91.0.4472.164
    • Edge (web)                  • edge          • web-javascript • Microsoft Edge 91.0.864.71

! Doctor found issues in 1 category.
```

```console
[✓] Flutter (Channel master, 2.4.0-5.0.pre.182, on Microsoft Windows [Version 10.0.19043.1110], locale en-US)
    • Flutter version 2.4.0-5.0.pre.182 at C:\Users\Taha\Code\flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ff49d35c90 (9 hours ago), 2021-07-26 02:28:03 -0400
    • Engine revision 420fa3a18f
    • Dart version 2.14.0 (build 2.14.0-350.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\android-sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_SDK_ROOT = C:\Users\Taha\Code\android-sdk
    • Java binary at: C:\Users\Taha\Code\android-studio-preview\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31515.178
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 4.2)
    • Android Studio at C:\Users\Taha\Code\android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[✓] Android Studio (version 2020.3)
    • Android Studio at C:\Users\Taha\Code\android-studio-preview
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.58.2)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.24.0

[✓] Connected device (5 available)
    • sdk google atv x86 (mobile) • emulator-5554 • android-x86     • Android 10 (API 29) (emulator)
    • Windows (desktop)           • windows       • windows-x64     • Microsoft Windows [Version 10.0.19043.1110]
    • Windows (UWP) (desktop)     • winuwp        • windows-uwp-x64 •
    • Chrome (web)                • chrome        • web-javascript  • Google Chrome 91.0.4472.164
    • Edge (web)                  • edge          • web-javascript  • Microsoft Edge 91.0.864.71

• No issues found!
```
		
</details>

✅: No Issue ❌: Issue reproduced

**Comment by predatorx7** [comment]
I have a non-breaking fix for this problem. Should I create a pull request? @TahaTesser

**Comment by xu-baolin** [comment]
> I have a non-breaking fix for this problem. Should I create a pull request? @TahaTesser

Go ahead!

**Comment by predatorx7** [comment]
This is fixed in 3.24.0
